### PR TITLE
feature: Update to use BUMP format in place of Merkle path binary

### PR DESCRIPTION
**Pull request by sirdeggen**
Update to use BUMP format in place of Merkle path binary, also update to use chainhash.Hash in the merkle tree calculation and use rather than reversing all the bytes into strings just to reverse them again to do other calculations - I hope much more efficient this way.

### CODE REVIEW

**Comment by sirdeggen** [comment]
Needs single path version, this is compound at the moment.

**Comment by sirdeggen** [comment]
The tests which are failing do not seem to be related to this PR, @shotasilagadzetaal please take a look when you have a moment.

**Comment by boecklim** [comment]
> The tests which are failing do not seem to be related to this PR, @shotasilagadzetaal please take a look when you have a moment.

@sirdeggen 
The tests are failing here: https://github.com/bitcoin-sv/arc/actions/runs/6737232351/job/18318287848?pr=121#step:11:26
--- FAIL: TestHandleBlock (0.00s)
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0

Ultimately the error happens here: https://github.com/libsv/go-bc/blob/58361c8e72a7a3c4259b2857a7f1c8a63afe7f3f/bump.go#L244

So I guess there is a bug in `go-bc`

**Comment by sirdeggen** [comment]
UPDATE: go-bc version updated based on the [fixes detailed here](https://github.com/libsv/go-bc/pull/75).

The test data in this repo includes a block with only one tx, the BUMP format had two bugs related to creating the correct struct and calculating the root in that case. Now resolved, all tests pass.

File: blocktx/peer_handler.go [hunk 1]
```diff
@@ -365,38 +353,32 @@ func (bs *PeerHandler) insertBlock(blockHash *chainhash.Hash, merkleRoot *chainh
...
-func (bs *PeerHandler) markTransactionsAsMined(blockId uint64, transactionHashes []*chainhash.Hash, merkleTree []string) error {
+func (bs *PeerHandler) markTransactionsAsMined(blockId uint64, merkleTree []*chainhash.Hash, blockHeight uint64) error {
 	start := gocore.CurrentNanos()
 	defer func() {
 		gocore.NewStat("blocktx").NewStat("HandleBlock").NewStat("markTransactionsAsMined").AddTime(start)
 	}()
 
 	txs := make([]*blocktx_api.TransactionAndSource, 0, TransactionStoringInterval)
 	merklePaths := make([]string, 0, TransactionStoringInterval)
+	transactionHashes := merkleTree[:(len(merkleTree)+1)/2]
```
Comment (shotasilagadzetaal) [inline blocktx/peer_handler.go:364]: Do we really have that many transactions in the block? What if number of transactions is not power of 2 ? 🤔  Correct me if I'm wrong but will this line be correct ?

Comment (sirdeggen) [reply blocktx/peer_handler.go]: Resolved, called them leaves.

File: blocktx/peer_handler.go [hunk 2]
```diff
@@ -365,39 +353,39 @@ func (bs *PeerHandler) insertBlock(blockHash *chainhash.Hash, merkleRoot *chainh
...
-func (bs *PeerHandler) markTransactionsAsMined(blockId uint64, transactionHashes []*chainhash.Hash, merkleTree []string) error {
+func (bs *PeerHandler) markTransactionsAsMined(blockId uint64, merkleTree []*chainhash.Hash, blockHeight uint64) error {
 	start := gocore.CurrentNanos()
 	defer func() {
 		gocore.NewStat("blocktx").NewStat("HandleBlock").NewStat("markTransactionsAsMined").AddTime(start)
 	}()
 
 	txs := make([]*blocktx_api.TransactionAndSource, 0, TransactionStoringInterval)
 	merklePaths := make([]string, 0, TransactionStoringInterval)
+	leaves := merkleTree[:(len(merkleTree)+1)/2]
```
Comment (sirdeggen) [inline blocktx/peer_handler.go:364]: Renamed to leaves.

File: blocktx/peer_handler.go [hunk 3]
```diff
@@ -365,39 +353,39 @@ func (bs *PeerHandler) insertBlock(blockHash *chainhash.Hash, merkleRoot *chainh
...
 
-	for txIndex, hash := range transactionHashes {
+	for txIndex, hash := range leaves {
+		// Everything to the right of the first nil will also be nil, as this is just padding upto the next PoT.
+		if hash == nil {
```
Comment (sirdeggen) [inline blocktx/peer_handler.go:368]: Avoids unnecessary work at the right hand side of the tree.

**Review by sirdeggen — COMMENTED** [review]
Updated based on comments from @shotasilagadzetaal, cheers.

File: metamorph/store/dynamodb/dynamodb.go [hunk 4]
```diff
@@ -469,4 +469,3 @@ func (ddb *DynamoDB) Close(ctx context.Context) error {
 	ctx.Done()
 	return nil
 }
-
```
Comment (sirdeggen) [inline metamorph/store/dynamodb/dynamodb.go:472]: gofmt -s -w . && git diff --exit-code

Made me do it.

**Review by shotasilagadzetaal — APPROVED** [review]
Great work

**Comment by sirdeggen** [comment]
The failure is related to an issue in go-bc found by Shota. Just noting here that will be fixed prior to merging.

Problem is somehow a nil is being passed to `MerkleTreeParentBytes` when it shouldn't be.

**Comment by sirdeggen** [comment]
Good to Merge from my perspective.

**Comment by sirdeggen** [comment]
CHANGELOG will be added prior to merge

File: CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
+# CHANGELOG
+
+All notable changes to this project will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## Table of Contents
+
+- [Unreleased](#unreleased)
+- [1.0.0 - YYYY-MM-DD](#100---yyyy-mm-dd)
```
Comment (boecklim) [inline CHANGELOG.md:8]: @sirdeggen 
Not sure what is the idea of this field, but shouldn't there be a date and the actual version of ARC?
I think the ARC version would be the tag, not?

**Review by nozim — APPROVED** [review]

File: blocktx/peer_handler.go [hunk 6]
```diff
@@ -365,39 +353,39 @@ func (bs *PeerHandler) insertBlock(blockHash *chainhash.Hash, merkleRoot *chainh
...
 		}
 
-		if root != merkleTree[len(merkleTree)-1] {
-			return errors.New("merkle path calculated produced different root")
-		}
-
-		merklePaths = append(merklePaths, merklePathBinary)
-		if txIndex%TransactionStoringInterval == 0 || txIndex == len(transactionHashes)-1 {
+		merklePaths = append(merklePaths, bumpHex)
+		if txIndex%TransactionStoringInterval == 0 || txIndex == len(leaves)-1 {
```
Comment (nozim) [inline blocktx/peer_handler.go:388]: @shotasilagadzetaal @boecklim can you please help clarify this line?

File: CHANGELOG.md (hunk 5 above)
Comment (shotasilagadzetaal) [reply CHANGELOG.md:8]: @boecklim  can we resolve or fix this somehow? Then we can merge

Comment (sirdeggen) [reply CHANGELOG.md:8]: The idea is to note releases here. We just added unreleased details here. You would add a date when there is a new release.

File: blocktx/peer_handler.go (hunk 6 above)
Comment (sirdeggen) [reply blocktx/peer_handler.go:388]: Noting that it's not my idea. I just renamed the variable.

**Pull request merged**